cache/in-memory/src/lib.rs

```diff
@@ -782,6 +814,29 @@ impl InMemoryCache {
         user
     }
 
+    fn cache_ref_user(&self, user: &User, guild_id: Option<GuildId>) -> Arc<User> {
+        match self.0.users.get_mut(&user.id) {
+            Some(mut u) if &*u.0 == user => {
+                if let Some(guild_id) = guild_id {
+                    u.1.insert(guild_id);
+                }
+
+                return Arc::clone(&u.value().0);
+            }
+            Some(_) | None => {}
+        }
+        let user = Arc::new(user.to_owned());
+        let mut guild_id_set = BTreeSet::new();
+        if let Some(guild_id) = guild_id {
+            guild_id_set.insert(guild_id);
+        }
+        self.0
+            .users
+            .insert(user.id, (Arc::clone(&user), guild_id_set));
```

Comment:
It's only worth doing this insert if there's actually a guild ID here, we don't need to allocate empty sets for future tracking